services/worker/src/worker/job_runners/config/parquet_and_info.py

```diff
@@ -0,0 +1,970 @@
+        if all(ref.ref != target_revision for ref in refs.converts):
+            committer_hf_api.create_branch(
+                repo_id=dataset, branch=target_revision, repo_type=DATASET_TYPE, revision=source_revision
+            )
+    except RepositoryNotFoundError as err:
+        raise DatasetNotFoundError("The dataset does not exist on the Hub.") from err
+
+    target_dataset_info = hf_api.dataset_info(repo_id=dataset, revision=target_revision, files_metadata=False)
+    # - check if there are files for configs that do not exist anymore and delete them
+    # - get configs that exist in repo
+    repo_parquet_files = {f.rfilename for f in target_dataset_info.siblings if f.rfilename.endswith(".parquet")}
+    repo_config_names = {f.split("/")[0] for f in repo_parquet_files}
+    delete_operations: List[CommitOperation] = []
+    legacy_configs = repo_config_names - config_names
+    # - delete configs that do not exist anymore
+    if legacy_configs:
+        legacy_configs_files_to_delete = {
+            file
+            for legacy_config in legacy_configs
+            for file in repo_parquet_files
+            if file.startswith(f"{legacy_config}/")
+        }
+        delete_operations.extend([CommitOperationDelete(path_in_repo=file) for file in legacy_configs_files_to_delete])
+        logging.debug(f"delete operations for legacy configs={delete_operations}")
+
+    # - delete existing files for current config
+    previous_config_repo_files = {filename for filename in repo_parquet_files if filename.startswith(f"{config}/")}
+    # - except for the files that we will update and .gitattributes if present.
+    config_files_to_add = {parquet_file.repo_file(): parquet_file.local_file for parquet_file in local_parquet_files}
+    config_files_to_delete = previous_config_repo_files - set(config_files_to_add).union({".gitattributes"})
+    delete_operations.extend([CommitOperationDelete(path_in_repo=file) for file in config_files_to_delete])
+    logging.debug(f"all delete operations={delete_operations}")
+
+    # send the files to the target revision
+    add_operations: List[CommitOperation] = [
+        CommitOperationAdd(path_in_repo=file, path_or_fileobj=local_file)
+        for (file, local_file) in config_files_to_add.items()
+    ]
+    logging.debug(f"{add_operations=}")
+
+    committer_hf_api.create_commit(
+        repo_id=dataset,
+        repo_type=DATASET_TYPE,
+        revision=target_revision,
+        operations=delete_operations + add_operations,
+        commit_message=commit_message,
+        parent_commit=target_dataset_info.sha,
+    )
+
+    # call the API again to get the list of parquet files
+    target_dataset_info = hf_api.dataset_info(repo_id=dataset, revision=target_revision, files_metadata=True)
+    repo_files = [
+        repo_file
+        for repo_file in target_dataset_info.siblings
+        if repo_file.rfilename.startswith(f"{config}/") and repo_file.rfilename.endswith(".parquet")
```

Comment:
Are we sure the directory is always `{config}/`?

If not, we should apply the same string normalization as done in the datasets library (help from other @huggingface/datasets is welcome)

It can be in another PR too, because the bug was already present in /parquet-and-dataset-info (but slightly different: here we will just ignore configs with weird names, while in /parquet-and-dataset-info: we include them, but with the `config` field in parquet file JSON being the config directory, not the config name)

Reply:
Hm from what I see in `ParquetFile.repo_files()` above which is copied from `/parquet-and-dataset-info`, it creates path which always starts from `"{config"/"`

`datasets.get_config_names` doesn't normalize configs and returns them just as they are specified by users or created automatically (for packaged builders like `csv`, `jsonl` etc.) so they can contain spaces for example and it might be a problem. Config names are only checked for [these symbols](https://github.com/huggingface/datasets/blob/1c4ec00511868bd881e84a6f7e0333648d833b8e/src/datasets/naming.py#L30): 
```python
INVALID_WINDOWS_CHARACTERS_IN_PATH = r"<>:/\|?*"
```
It would raise at dataset creation so we won't get configs with these symbols for sure. But they are not normalized after.
Am I right? cc @albertvillanova @lhoestq 

> here we will just ignore configs with weird names, while in /parquet-and-dataset-info: we include them, but with the config field in parquet file JSON being the config directory, not the config name

I think I don't understand what this mean... Where these differences are reflected in the code? I think I'm missing something

Reply:
> they can contain spaces for example and it might be a problem

 but it's normalized by HF API.

Reply:
> > here we will just ignore configs with weird names, while in /parquet-and-dataset-info: we include them, but with the config field in parquet file JSON being the config directory, not the config name
> I think I don't understand what this mean... Where these differences are reflected in the code? I think I'm missing something

In config-parquet-and-info, here: if a config has a weird name, it will not be included in the list of parquet files in the response

In /parquet-and-dataset-info, if a config has a weird name, it is included in the list of parquet files in the response, but the "config" field is the name of the directory, not necessarily the name of the config, if they are different due to weird characters

Reply:
OK, thanks for the research. So, we surely want to delegate the comparison and/or conversion to specialized functions (so that we can unit test, and have all the logic in one place).